qt-5: add buildPackages.stdenv.cc to nativeBuildInputs

amjoseph-nixpkgs opened the pull request:
###### Description of changes

Because `qt-5` uses `lib.newScope` rather than `lib.newScopeWithSplicing`, the splicing mechanism for cross compilation does not appear to work correctly.

Rather than risk breakage on native compiles, let's just choose the `buildPackages` version manually for `perl` and the host `cc` (which is needed for `qmake`).  This commit does that by adding them to `nativeBuildInputs` in `qtModule.nix` and `modules/qtbase.nix`.

Commit:
qt-5: add buildPackages.stdenv.cc to nativeBuildInputs

Qmake expects to be able to find the host's C compiler in the $PATH
when cross compiling.  Let's add it.

Co-authored-by: superherointj <5861043+superherointj@users.noreply.github.com>

superherointj commented on pkgs/development/libraries/qt-5/modules/qtbase.nix:
```diff
@@ -77,7 +79,8 @@ stdenv.mkDerivation (finalAttrs: {
   nativeBuildInputs = [ bison flex gperf lndir perl pkg-config which ]
-    ++ lib.optionals stdenv.isDarwin [ xcbuild ];
+    ++ lib.optionals stdenv.isDarwin [ xcbuild ]
+    ++ lib.optionals (stdenv.hostPlatform != stdenv.buildPlatform) buildPackages.stdenv.cc ];
```
```suggestion
    ++ lib.optional (stdenv.hostPlatform != stdenv.buildPlatform) buildPackages.stdenv.cc;
```

superherointj replied:
I have no opinion on `lib.optional` vs `lib.optionals`. Issue here is the bracket that never started. As it is a single use, I usually use `lib.optional`. If remember there was a debate once related to this but I don't know if consensus was reached.

superherointj commented on pkgs/development/libraries/qt-5/qtModule.nix:
```diff
@@ -16,7 +20,9 @@ mkDerivation (args // {
-  nativeBuildInputs = (args.nativeBuildInputs or []) ++ [ perl self.qmake ];
+  nativeBuildInputs = (args.nativeBuildInputs or [])
+                      ++ [ buildPackages.perl self.qmake ]
+                      ++ lib.optionals (stdenv.hostPlatform != stdenv.buildPlatform) buildPackages.stdenv.cc ];
```
Same bracket problem here. Won't eval as it is.

Artturin commented:
this isn't related to newScope, instead the issue is that the inherits in all-packages aren't inherited from `__splicedPackages`

change
```nix
qt5 = recurseIntoAttrs (makeOverridable
  (import ../development/libraries/qt-5/5.15) {
    inherit newScope;
    inherit lib fetchurl fetchpatch fetchgit fetchFromGitHub makeSetupHook makeWrapper;
    inherit bison cups dconf harfbuzz libGL perl gtk3 python3;
    inherit (gst_all_1) gstreamer gst-plugins-base;
    inherit darwin;
    inherit buildPackages;
    stdenv = if stdenv.isDarwin then darwin.apple_sdk_11_0.stdenv else stdenv;
  });
```
to
```nix
  qt5 = recurseIntoAttrs (makeOverridable
    (import ../development/libraries/qt-5/5.15) {
      inherit (__splicedPackages)
        newScope lib fetchurl fetchpatch fetchgit fetchFromGitHub makeSetupHook makeWrapper
        bison cups dconf harfbuzz libGL perl gtk3 python3
        darwin buildPackages;
      inherit (__splicedPackages.gst_all_1) gstreamer gst-plugins-base;
      stdenv = if stdenv.isDarwin then darwin.apple_sdk_11_0.stdenv else stdenv;
    });
```

Artturin commented:
https://github.com/NixOS/nixpkgs/pull/220374 inherit  __splicedPackages + makeScopeWithSplicing

amjoseph-nixpkgs commented:
Rebased on top of https://github.com/NixOS/nixpkgs/pull/220374

This allowed to drop the `buildPackages.` prefix from `perl`.  The addition of `buildPackages.stdenv.cc` is still necessary however.

Artturin commented:
adding buildPackages.stdenv.cc to nativeBuildInputs looks wrong, investigating..

Artturin commented:
qmake/Makefile (which is generated) has the wrong CC and CXX variables

also for ref here's the qt5base build file for build root
https://github.com/buildroot/buildroot/blob/master/package/qt5/qt5base/qt5base.mk
i recommend cloning that repository
here's a couple more cross focused repos
https://github.com/openembedded/openembedded-core
https://git.yoctoproject.org/poky

amjoseph-nixpkgs commented:
> qmake/Makefile (which is generated) has the wrong CC and CXX variables (non prefixed binaries)

Is that wrong? `qmake` runs on the `buildPlatform` (counterintuitively).

In theory in order to fit nixpkgs way of thinking, `qmake` ought to be in its own derivation separate from `qtbase`.  But that would be a pretty huge undertaking for a library that is no longer actively developed upstream...

amjoseph-nixpkgs commented:
> i recommend cloning that repository here's a couple more cross focused repos

Void is another good one.

Artturin commented:
> > qmake/Makefile (which is generated) has the wrong CC and CXX variables (non prefixed binaries)
> 
> Is that wrong? `qmake` runs on the `buildPlatform` (counterintuitively).
> 
> In theory in order to fit nixpkgs way of thinking, `qmake` ought to be in its own derivation separate from `qtbase`. But that would be a pretty huge undertaking for a library that is no longer actively developed upstream...
> 
> Edit: or we could think of `qmake` as being a compiler, and give it a `targetPlatform` (so `qmake`'s `(build==host)!=target`)

`pkgsCross.aarch64-multiplatform.buildPackages.qt5.qtbase` (nativeBuildInputs) already builds, while this is making it possible to build `pkgsCross.aarch64-multiplatform.qt5.qtbase`(buildInputs) from which binaries won't be run on buildPlatform so using the non-prefixd binaries to build it seems like a bug

SuperSandro2000 commented on pkgs/development/libraries/qt-5/5.15/default.nix:
```diff
@@ -106,7 +106,7 @@ let
         import ../qtModule.nix
         {
           inherit perl;
```
```suggestion
```

SuperSandro2000 commented on pkgs/development/libraries/qt-5/5.15/default.nix:
```diff
@@ -106,7 +106,7 @@ let
         {
           inherit perl;
-          inherit lib;
+          inherit lib stdenv buildPackages;
```
```suggestion
          inherit perl lib stdenv buildPackages;
```

SuperSandro2000 commented on pkgs/development/libraries/qt-5/modules/qtbase.nix:
```diff
@@ -77,7 +79,11 @@ stdenv.mkDerivation (finalAttrs: {
   nativeBuildInputs = [ bison flex gperf lndir perl pkg-config which ]
-    ++ lib.optionals stdenv.isDarwin [ xcbuild ];
+    ++ lib.optionals stdenv.isDarwin [ xcbuild ]
+    # `qtbase` expects to find `cc` (with no prefix) in the
```
Can we patch that?

Why do we need this in the wrapper and package?

amjoseph-nixpkgs replied:
I agree, there's a lot of ugliness in nixpkgs' `qt5` expressions.  But I am trying really hard not to undertake a rewrite-from-scratch of them here.

SuperSandro2000 commented on pkgs/development/libraries/qt-5/qtModule.nix:
```diff
@@ -16,7 +21,12 @@ mkDerivation (args // {
   patches = (args.patches or []) ++ (patches.${pname} or []);
 
-  nativeBuildInputs = (args.nativeBuildInputs or []) ++ [ perl self.qmake ];
+  nativeBuildInputs = (args.nativeBuildInputs or [])
+                      ++ [ perl self.qmake ]
+                      # `qmake` expects to find `cc` (with no prefix) in the
+                      # `$PATH`, so the following is needed even if
+                      # `stdenv.buildPlatform.canExecute stdenv.hostPlatform`
+                      ++ lib.optional (stdenv.hostPlatform != stdenv.buildPlatform) buildPackages.stdenv.cc;
```
```suggestion
  nativeBuildInputs = (args.nativeBuildInputs or [])
    ++ [ perl self.qmake ]
    # `qmake` expects to find `cc` (with no prefix) in the
    # `$PATH`, so the following is needed even if
    # `stdenv.buildPlatform.canExecute stdenv.hostPlatform`
    ++ lib.optional (stdenv.hostPlatform != stdenv.buildPlatform) buildPackages.stdenv.cc;
```

Commit:
qt5: move buildPackages.stdenv.cc to pkgsBuildBuild

As recommended here:
https://github.com/NixOS/nixpkgs/pull/220310#issuecomment-1464158888

The non-idiomatic `if..then..else null` is to prevent a mass-rebuild
sending this to staging.  I can submit a separate PR to staging to
change it to the `lib.optional` idiom if this bothers people.

Commit:
Update pkgs/development/libraries/qt-5/5.15/default.nix

Co-authored-by: Sandro <sandro.jaeckel@gmail.com>

Commit:
Update pkgs/development/libraries/qt-5/5.15/default.nix

Co-authored-by: Sandro <sandro.jaeckel@gmail.com>

Commit:
Update pkgs/development/libraries/qt-5/qtModule.nix

Co-authored-by: Sandro <sandro.jaeckel@gmail.com>

amjoseph-nixpkgs commented:
> `pkgsCross.aarch64-multiplatform.buildPackages.qt5.qtbase` (nativeBuildInputs) already builds, while this is making it possible to build `pkgsCross.aarch64-multiplatform.qt5.qtbase`(buildInputs) from which binaries won't be run on buildPlatform so using the non-prefixd binaries to build it seems like a bug

I'm having trouble understanding this, except for the last bit about prefixes, which I disagree with: the fact that native compilers are unprefixed is a bug in nixpkgs (which is painful to fix) not a feature.

> adding `buildPackages.stdenv.cc` to `qtModule` is especially wrong
> 
> even if a build build compiler was needed then `buildPackages.stdenv.cc` should go in to depsBuildBuild

Since `qtbase` has no meaningful `targetPlatform` that's a bit of a nitpick, but okay: ceb4d990d2b8d3f9ef82333e6ab5b51022558e41
